### PR TITLE
HDDS-5432. Enable downgrade testing after 1.1.0 release.

### PR DESCRIPTION
**Pull request by errose28**
## What changes were proposed in this pull request?

Enable downgrade testing from current version of Ozone to 1.1.0. This involves implementing the volume formatting changes needed on the datanode in a backwards compatible way. For details, please see the design document that is attached to the Jira.

~The following tasks are remaining~
- [x] Descriptions of how corner cases have been handled to aid in reviewing.
    - Updated in the design doc attached to the Jira.
- [x] Unit testing of corner cases.
- [x] Fix Interrupt on datanode shutdown that is failing some integration tests in TestHDDSUpgrade.
- [x] Breakdown of changes to aid in reviewing.
- [x] HDDS-5503 ([PR](https://github.com/apache/ozone/pull/2476)) gets merged so the on finalize action does not need to be hard coded.

## What is the link to the Apache JIRA

HDDS-5432

## How was this patch tested?

- Downgrade acceptance testing enabled.
- Unit testing for corner cases in `TestDatanodeUpgradeToScmHA`.
    - `testReadsDuringFinalization` and `testImportContainer` passed 100 times each in local runs to verify they are not flaky.

## Breakdown of changes

- Although a lot of files were changed, only a few files deal with the onboarding of SCM HA as discussed in the design document:
    - `AbstractLayoutVersionManager`
        - Read-write locking added since metadata layout version can now be queried by closed container import during finalization.
   - `HddsVolumeUtil#checkVolume`
       - For handling formatting of new volumes.
  - `KeyValueContainer#create`
      - For handling new containers created.
  - `KeyValueHandler#populateContainerPathFields`
      - For setting absolute paths to chunk and metadata in imported container files.
  - `ContainerReader#readVolume`
      - Remove the dependency on a single cluster ID directory inside the *\<volume>/hdds* directory.
  - `ScmHAFinalizeUpgradeActionDatanode`
      - The action that adds the symbolic link from the cluster ID directory to the SCM ID directory on finalization of the datanode.
  - `VersionedDatanodeFeatures`
      - Utility class that is used to look up values to use based on layout versions.
      - This class contains the implementations of the `chooseContainerPathID` functions defined in the design document.

- The remaining changes are fairly minor and mostly fall under the following categories:
    - Removing the original backwards incompatible cluster ID directory reformatting code that was used before the upgrade framework.
    - Enabling commented out sections of the upgrade/downgrade acceptance tests that now work.
    - Removing the manual upgrade acceptance tests that no longer work for the current code which uses the upgrade framework for disk layout changes.
    - Removing old references to SCM ID usage in variable names and comments.
    - Rename/refactor of the old `DatanodeMetadataFeatures` class into the more robust and well named `VersionedDatanodeFeatures`

### CODE REVIEW

**Comment by errose28** [comment]
CI failure passed locally and seems unrelated. Retriggering.

File: hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/container/common/statemachine/DatanodeStateMachine.java [hunk 1]
```diff
@@ -452,8 +452,8 @@ public boolean isTransitionAllowedTo(DatanodeStates newState) {
   public void startDaemon() {
     Runnable startStateMachineTask = () -> {
       try {
-        start();
         LOG.info("Ozone container server started.");
+        start();
```
Comment (adoroszlai) [inline hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/container/common/statemachine/DatanodeStateMachine.java:456]: Great, datanodes will no longer print "server started" right when being stopped. :)  This has always bothered me, but not enough to fix it.

File: hadoop-hdds/common/src/main/java/org/apache/hadoop/ozone/upgrade/AbstractLayoutVersionManager.java [hunk 2]
```diff
@@ -50,9 +51,13 @@
...
   protected Map<String, T> featureMap = new HashMap<>();
   protected volatile Status currentUpgradeState = FINALIZATION_REQUIRED;
+  // Allows querying upgrade state while an upgrade is in progress.
+  // Note that MLV may have been incremented during the upgrade
+  // by the time the value is read/used.
+  private ReentrantReadWriteLock lock;
 
   protected void init(int version, T[] lfs) throws IOException {
-
+    lock = new ReentrantReadWriteLock();
```
Comment (adoroszlai) [inline hadoop-hdds/common/src/main/java/org/apache/hadoop/ozone/upgrade/AbstractLayoutVersionManager.java:60]: Nit: `lock` could be `final` and initialized at the declaration.  And `init()` may be guarded by the write lock.

File: hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/container/common/utils/HddsVolumeUtil.java [hunk 3]
```diff
@@ -168,18 +169,25 @@ public static String getProperty(
...
-  public static boolean checkVolume(HddsVolume hddsVolume, String scmId, String
-      clusterId, Logger logger) {
+  public static boolean checkVolume(HddsVolume hddsVolume, String scmId,
+      String clusterId, ConfigurationSource conf, Logger logger) {
     File hddsRoot = hddsVolume.getHddsRootDir();
     String volumeRoot = hddsRoot.getPath();
     File clusterDir = new File(hddsRoot, clusterId);
-    File scmDir = new File(hddsRoot, scmId);
+    String errorPrefix = "Volume " + volumeRoot + " is in an inconsistent " +
+        "state.";
```
Comment (adoroszlai) [inline hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/container/common/utils/HddsVolumeUtil.java:190]: Nit: This is used only in one of the branches.  Should it be defined only there?

File: hadoop-ozone/dist/src/main/compose/upgrade/test.sh [hunk 4]
```diff
@@ -32,12 +32,7 @@ run_test_scripts ${tests} || RESULT=$?
 RESULT_DIR="$ALL_RESULT_DIR" create_results_dir
 
 # Upgrade tests to be run.
-# Run all upgrades even if one fails.
-# Any failure will save a failing return code to $RESULT.
-set +e
-run_test manual-upgrade 0.5.0 1.1.0
-run_test non-rolling-upgrade 1.0.0 1.1.0
```
Comment (adoroszlai) [inline hadoop-ozone/dist/src/main/compose/upgrade/test.sh:39]: I didn't see these tests (and old version-specific logic) being removed mentioned in the PR.  I'm fine with it, but maybe the README should be updated accordingly.  Specifically: _... testing upgrades from any previous version to another previous version_ is no longer true (if I understand it correctly).

**Review by adoroszlai — APPROVED** [review]
Thanks @errose28 for the patch, it looks great.  I also like the bunch of small improvements included.  Maybe the patch would have been easier to digest if those were submitted separately, but I understand that it's a hassle to extract them.

**Review by errose28 — COMMENTED** [review]
Thanks for taking the time to review this large PR @adoroszlai. All your suggestions should be implemented in the latest set of changes.

File: hadoop-ozone/dist/src/main/compose/upgrade/test.sh (hunk 4 above)
Comment (errose28) [reply hadoop-ozone/dist/src/main/compose/upgrade/test.sh:39]: 1.0.0 -> 1.1.0 did not require any manual intervention. So the upgrade framework supports upgrading from 1.0.0 (GA) to any later version.

**Review by avijayanhwx — APPROVED** [review]
Thanks for working on this @errose28!

**Comment by errose28** [comment]
Thanks for taking the time to review and approve this large PR @adoroszlai @avijayanhwx. Merging it now.

**Comment by weimingdiit** [comment]
@errose28  Hi, the ReentrantReadWriteLock added to the AbstractLayoutVersionManager class will cause excessive CPU load. Please pay attention to this issue.  https://issues.apache.org/jira/browse/HDDS-11240

**Pull request merged**